### PR TITLE
btf: fix CO-RE relocations for local type id

### PR DESCRIPTION
**Pull request by lmb**
At some point the kernel gained the ability to dynamically allocated typed memory via a bpf_obj_new helper (see [1]). This helper receives the ID of the type to allocate as an argument.

Trying to use this helper with the library will currently fail, since we took a short-cut when implementing on the fly BTF generation. At that point in time there was no use of the local_type_id relocation in the kernel, so we just stuck with the value present in ext_infos. This is of course not correct when building a BTF blob from scratch: type IDs are reallocated, and only used types are passed to the kernel in the first place.

Fix this by allocating an ID for any type that is the target of a local_type_id relocation.

1: https://lwn.net/Articles/915404/

### CODE REVIEW

File: btf/core.go [hunk 1]
```diff
@@ -164,7 +164,7 @@ func (k coreKind) String() string {
 //
 // Fixups are returned in the order of relos, e.g. fixup[i] is the solution
 // for relos[i].
-func CORERelocate(relos []*CORERelocation, target *Spec, bo binary.ByteOrder) ([]COREFixup, error) {
+func CORERelocate(relos []*CORERelocation, target *Spec, bo binary.ByteOrder, typeID func(Type) (TypeID, error)) ([]COREFixup, error) {
```
Comment (ti-mo) [inline btf/core.go:167]: Would it be wrong to make CORERelocate take a Builder here? I understand this way the changes are kept minimal, but it doesn't convey clearly what (potentially) happens behind the scenes, being adding a type to the BTF blob. AFAICS, we only pass `spec.TypeID()` here in tests.

File: prog.go [hunk 2]
```diff
@@ -242,14 +242,24 @@ func newProgramWithOptions(spec *ProgramSpec, opts ProgramOptions) (*Program, er
...
+	var b btf.Builder
+	if err := applyRelocations(insns, opts.KernelTypes, spec.ByteOrder, &b); err != nil {
+		return nil, fmt.Errorf("apply CO-RE relocations: %w", err)
 	}
-	if handle != nil {
-		defer handle.Close()
 
-		attr.ProgBtfFd = uint32(handle.FD())
+	if !b.Empty() && haveProgramExtInfos() != nil {
+		// Return ErrNotSupported instead of E2BIG if CO-RE relies on BTF
```
Comment (ti-mo) [inline prog.go:251]: It's not clear what you mean by this. Doesn't CO-RE always rely on BTF being present?

File: prog.go [hunk 3]
```diff
@@ -242,14 +242,24 @@ func newProgramWithOptions(spec *ProgramSpec, opts ProgramOptions) (*Program, er
...
+		// being present.
+		return nil, haveProgramExtInfos()
```
Comment (ti-mo) [inline prog.go:253]: `haveProgramExtInfos()` is not guaranteed to return `ErrNotSupported` here. I'd either make the surrounding if contain `errors.Is(haveProgramExtInfos(), ErrNotSupported)` or `return nil, ErrNotSupported` here.

File: prog.go [hunk 4]
```diff
@@ -242,14 +242,24 @@ func newProgramWithOptions(spec *ProgramSpec, opts ProgramOptions) (*Program, er
...
+	}
+
+	if haveProgramExtInfos() == nil {
```
Comment (ti-mo) [inline prog.go:256]: Maybe it's time to move this to a function? `newProgram` is getting crowded.

**Review by ti-mo — APPROVED** [review]
Thanks! Few questions/nits.

File: prog.go (hunk 2 above)
Comment (lmb) [reply prog.go]: I'll rephrase. The issue is that a CO-RE local type id relocation might assert that a specific type exists in the BTF blob.

File: prog.go (hunk 3 above)
Comment (lmb) [reply prog.go]: Good point.

File: prog.go (hunk 4 above)
Comment (lmb) [reply prog.go]: I'll take another look!

Comment (lmb) [reply prog.go]: Yes, it needs a refactor. But I don't have the time at the moment.

**Pull request merged**